### ISSUE TITLE
transform-es2015-modules-systemjs breaks ES6 class definition

### ISSUE THREAD

**Issue by freshp86** [babel/babel#4701]
Original file

```
// foo_module.js original
class Foo {
  constructor() { this.foo = 'foo'; }
}

export function createInstance() {
  return new Foo();
}
```

After passing through the 'transform-es2015-modules-systemjs' plugin, the class definition is incorrectly moved within the "execute" section. See below

```
// foo_module.js after systemjs plugin
System.register([], function (_export, _context) {
  "use strict";

  function createInstance() {
    return new Foo();
  }

  _export('createInstance', createInstance);

  return {
    setters: [],
    execute: function () {
      // Class definition should not have been moved here. It is no longer visible by
      // the createInstance() function.
      class Foo {
        constructor() {
          this.foo = 'foo';
        }
      }
    }
  };
});
```

Now when I try to import the module I get an error

```
Uncaught (in promise) ReferenceError: Foo is not defined(..)
```

For example I use this HTML file

```
<html>
  <head>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/systemjs/0.19.39/system.js"></script>
  </head>
  <body>
    <script>
      System.import('./foo_module.js').then(function(m) {
        m.createInstance(); // Throws error
      });
    </script>
  <body>
</html>
```

If I change the original file to define the class using the pre-ES6 way, as follows

```
function Foo() { this.foo = 'foo'; }
```

it works fine, because the plugin does not move anything in the 'execute' section, see below

```
System.register([], function (_export, _context) {
  "use strict";

  function Foo() {
    this.foo = 'foo';
  }

  function createInstance() {
    return new Foo();
  }

  _export('createInstance', createInstance);

  return {
    setters: [],
    execute: function () {}
  };
});
```

**Comment by DrewML** [comment]
@freshp86 Thanks for the report! It looks like this broke in July with 83522c9 in PR https://github.com/babel/babel/pull/3594. I've verified by going to 253918c locally that the issue is no longer present.

Output with 253918c:

```javascript
System.register([], function (_export, _context) {
  "use strict";

  return {
    setters: [],
    execute: function () {
      class Foo {
        constructor() {
          this.foo = 'foo';
        }
      }

      function createInstance() {
        return new Foo();
      }

      _export('createInstance', createInstance);
    }
  };
});
```

We can't revert that change, though, since it has other necessary fixes in it.

Tagging @asapach just in case he's willing to help out 😃

**Comment by asapach** [comment]
> After passing through the 'transform-es2015-modules-systemjs' plugin, the class definition is incorrectly moved within the "execute" section.

It's the other way around - class declaration is not hoisted from the `execute` function, unlike functions.
> it works fine, because the plugin does not move anything in the 'execute' section, see below

Again, incorrect assumption: all the functions are hoisted, so the `execute` section becomes empty.

We could just hoist all the class declarations similar to functions, but the spec says they are not to be hoisted: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes#Hoisting, so it would change the behavior.

Not sure how to handle this properly.

**Comment by DrewML** [comment]
Did a bit more reading on this (wasn't familiar with the specifics of `System.register`). For reference, [current documentation is here](https://github.com/ModuleLoader/es-module-loader/blob/v0.17.0/docs/system-register.md).

The text addresses how to handle things that _should_ be hoisted (functions + variables), but it doesn't appear to specify (at least, based on my understanding) how to handle this use-case.

My first thought was that we'd need to transform to something like this (basically what @asapach alluded to):

```javascript
System.register([], function (_export, _context) {
  "use strict";
  var Foo;

  function createInstance() {
    return new Foo();
  }

  _export('createInstance', createInstance);

  return {
    setters: [],
    execute: function () {
      Foo = class Foo {
        constructor() {
          this.foo = 'foo';
        }
      }
    }
  };
});
```

But that wouldn't be spec compliant, because that would cause code like the following (contrived, of course) to behave incorrectly

```javascript
export function createInstance() {
   return new Foo();
}

createInstance(); // should throw a ReferenceError, but instead would throw a TypeError, since Foo would be undefined

class Foo {
   constructor() {}
}
```

:thinking:

**Comment by freshp86** [comment]
FWIW, TypeScript compiler with --module=system seems to handle the initially reported bug correctly. So perhaps try to replicate TS compiler's output?

**Comment by DrewML** [comment]
@freshp86 Good call! I had just started looking for another compiler that output system modules.

For the input

```javascript
class Foo {
  constructor() {  }
}

export function createInstance() {
  return new Foo();
}
```

TypeScript, using `tsc index.ts --system module --target es6` produces:

```javascript
System.register([], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var Foo;
    function createInstance() {
        return new Foo();
    }
    exports_1("createInstance", createInstance);
    return {
        setters:[],
        execute: function() {
            Foo = class Foo {
                constructor() {
                }
            };
        }
    }
});
```

Which certainly works, but I believe that still has the same problem I referenced above.

**Input**:
```javascript
export function createInstance() {
  return new Foo();
}

createInstance();

class Foo {
  constructor() {  }
}
```

**Output**:
```javascript
System.register([], function(exports_1, context_1) {
    "use strict";
    var __moduleName = context_1 && context_1.id;
    var Foo;
    function createInstance() {
        return new Foo();
    }
    exports_1("createInstance", createInstance);
    return {
        setters:[],
        execute: function() {
            createInstance();
            Foo = class Foo {
                constructor() {
                }
            };
        }
    }
});
```

I suppose, for now, we could at least get parity with how TS is handling things.

**Comment by DrewML** [comment]
I have a fix for this locally, that also takes care of the bug I listed above that can happen with the TS version. Need to fix one thing with it, and then I'll open a PR.

**Comment by loganfsmyth** [comment]
The `var` approach seems fine to me. SystemJS's format doesn't provide any way I could see to preserve TDZ behavior, and it'd be tough/impossible to support depending on your target environment anyway.

**Comment by nicolo-ribaudo** [comment]
This is fixed in Babel 7.4 (https://github.com/babel/babel/pull/9639)
